core/services/data_source.go

```diff
@@ -57,26 +59,34 @@ func NewDataSources(id string, trigger *webhook.Trigger, runChannel *chan *big.I
 // and return the job run result here as a "proxy" to the EAs
 // this allows the CL node to track OCR rounds/job runs
 func (dss *dataSourceState) Observe(ctx context.Context) error {
-	// https://stackoverflow.com/a/45211470
-	// use low level memory utility to perform locking/unlocking
-	if !atomic.CompareAndSwapUint32(&dss.locked, 0, 1) {
+	// lock
+	if !dss.locked.CAS(false, true) {
 		return errAlreadyTriggered
 	}
-	defer atomic.StoreUint32(&dss.locked, 0)
+	defer dss.locked.Store(false)
 
 	// set observation request started
 	dss.Log.Infof("[%s] Observe (job run) triggered", dss.ID)
 	dss.Done = make(chan struct{}) // create channel for signaling finish
+	defer close(dss.Done)          // close channel to indicate done
 
 	// send job trigger
 	go dss.Webhook.TriggerJob(dss.ID)
 
-	// wait for job run data to be returned
-	dss.Prices[priceFeedParam] = <-*dss.RunData // use first value
-	dss.Prices[juelsToXParam] = <-*dss.RunData  // use second value
+	// wait for job run data to be returned (each channel read is wrapped with ctx.Done())
+	select {
+	case dss.Prices[priceFeedParam] = <-*dss.RunData: // use first value
+	case <-ctx.Done():
+		return fmt.Errorf("[%s] Observe (job run) cancelled", dss.ID)
+	}
+	select {
+	case dss.Prices[juelsToXParam] = <-*dss.RunData: // use second value
+	case <-ctx.Done():
+		return fmt.Errorf("[%s] Observe (job run) cancelled", dss.ID)
+	}
+
```

Comment:
Sending back two separate sequential values and treating them differently feels kind of weird. What if they arrive out of order? If the same routine is sending them both back, just use a `struct x { priceFeed, juelsToUnit }`

Reply:
(Also, how does `core` handle this?)

Reply:
this probably not be relevant anymore because of switching over to core but i'll answer them 😄 
* the order should be fine since there's a part that deconstructs a struct and returns two separate values: https://github.com/smartcontractkit/chainlink-relay/blob/7839e322d0e1e6022f266c6b3b9425556a97404f/core/services/ocr2.go#L186 (and yes, if we continue to use this i should just return a struct rather than values twice)
* it's just mocked in core right now 😆